src/libstd/collections/hash/map.rs

```diff
@@ -1195,13 +630,28 @@ impl<K, V, S> HashMap<K, V, S>
     /// assert_eq!(letters[&'u'], 1);
     /// assert_eq!(letters.get(&'y'), None);
     /// ```
+    #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn entry(&mut self, key: K) -> Entry<K, V> {
-        // Gotta resize now.
+        // Ideally we would put this in VacantEntry::insert, but Entry is not
+        // generic over the BuildHasher and adding a generic parameter would be
+        // a breaking change.
```

Comment:
Because the reserve could invalidate the result of the lookup if it rehashes the table.

Reply:
It is ok if it does. In most cases there is no need to rehash. If a rehash happens, the cost of an additional probe round is insignificant. Right now this reserve(1) is breaking the API guarantees. `reserve(N)` followed by N `entry(K).insert(V)` will cause a rehash even though it is specified that it will not.

Reply:
Sorry, you were right, this reserve call should indeed be in the `else` block.

We can't defer the reserve call until the entry is inserted because `Entry` would need to hold a reference to the hasher, which is not possible since in libstd `Entry` is `Entry<K, V>` instead of `Entry<K, V, S>` and changing that would be a breaking change.